File: src/boj/boj_1874.java
package boj;

import java.io.*;
import java.util.Stack;

public class boj_1874 {
    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        final StringBuilder stringBuilder = new StringBuilder();
        final Stack<Integer> stack = new Stack<>();
        final int n = Integer.parseInt(br.readLine());

        int num = 0;
        for (int i = 0; i < n; i++) {
            final int target = Integer.parseInt(br.readLine());
            if (num <= target) {
                while (num != target) {
                    stack.push(++num);
                    stringBuilder.append("+\n");
                }
                stack.pop();
            } else {
                final int pop = stack.pop();
                if (pop != target) {
                    System.out.println("NO");
                    return;
                }
            }
            stringBuilder.append("-\n");
        }
        System.out.print(stringBuilder);
    }
}

File: src/boj/boj_1874.kt
package boj

import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val answer = LinkedList<Char>()
    val stack = LinkedList<Int>()
    val n = readLine().toInt()
    var i = 1
    for (j in 0 until n) {
        val m = readLine().toInt()
        if (i <= m) {
            while (i <= m) {
                stack.addLast(i++)
                answer.addLast('+')
            }
            stack.pollLast()
            answer.addLast('-')
        } else {
            do {
                val pop = stack.pollLast()
                if (pop != m) {
                    println("NO")
                    return@with
                }
                answer.addLast('-')
            } while (pop != m)
        }
    }

    val bufferedWriter = System.out.bufferedWriter()
    answer.forEach {
        bufferedWriter.append("$it\n")
    }
    bufferedWriter.flush()
    bufferedWriter.close()
}